### PR TITLE
Update comparison logic for worker state

### PR DESCRIPTION
**Pull request by rockwellw**
Addresses #3256 , based on @TomAugspurger 's comment.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks. Do you think this will actually fix #3256? This being the cause was a shot in the dark.

And do you know if / how `WorkerState.__eq__` is defined? I don't see one at a glance. Does implementing `__slots__` give us one?

**Comment by rockwellw** [comment]
This may just address a symptom and not the root problem, but it seems like a good starting point considering the log messages @bnaul was seeing.

**Comment by TomAugspurger** [comment]
Ok. Do you know how `__eq__` is implemented ok WorkerState?

**Comment by TomAugspurger** [comment]
I think `WorkerState.__eq__` may actually be using identity. I would have expected these to be equal, if we were comparing attributes

```python
In [15]: from distributed.scheduler import WorkerState

In [16]: a = WorkerState()

In [17]: b = WorkerState()

In [18]: a == b
Out[18]: False
```

This may not be a problem if WorkerState instances are expected to be singletons per worker, which I suspect they are.

**Comment by StephanErb** [comment]
There is another occurrence of the same identity comparison here https://github.com/dask/distributed/blob/d747f639d73c5f9b8961f2f80e407868de7da4cb/distributed/scheduler.py#L2560 To reduce surprises, I guess both should be fixed.

While I have no proof, I could totally imagine that race conditions between `add_worker` and `remove_worker` break the single instance assumption.

**Comment by bnaul** [comment]
> I think `WorkerState.__eq__` may actually be using identity. I would have expected these to be equal, if we were comparing attributes

@TomAugspurger it looks like `identity()` has the info that we would want to compare against, do you think something like this might do the trick?
```
class WorkerState():
    def __eq__(self, value):
        if hasattr(value, 'identity'):
            return self.identity() == value.identity()
        else:
            return False
```

**Comment by mrocklin** [comment]
Rather than conditionals I usually do something like ...

    return type(self) == type(other) and self.identity() == other.identity()

Also, if we're going to implement `__eq__` then we'll probably also have to implement `__hash__` (should be more or less the same)

**Comment by bnaul** [comment]
> Rather than conditionals I usually do something like ...
> 
> ```
> return type(self) == type(other) and self.identity() == other.identity()
> ```
Probably not at all important for this case but doesn't this pattern kind of break subclassing? How about `isinstance(other, type(self))`?

> Also, if we're going to implement `__eq__` then we'll probably also have to implement `__hash__` (should be more or less the same)
Do we currently rely on `WorkerState`s being hashable? It seems to me like they're mutable so wouldn't we want `hash` to be `None`?

**Comment by mrocklin** [comment]
If they're of different classes then can they be equal?  (but yeah, in truth it doesn't matter)

Fair enough on hashability (assuming that it doesn't come up, which it may)

**Comment by bnaul** [comment]
@mrocklin you were right that hashability is needed so we added that as well and changed `__eq__` to rely on that instead. it seems to me like just name+host is enough to uniquely specify the WorkerState but if you think we should add more attributes we can do that.

@TomAugspurger anecdotally we haven't seen #3256 since we started running with this change, but I wouldn't say we've conclusively proven it's resolved since unfortunately that's a random unpredictable failure. regardless this seems like the "right" check to be performing and should help w/ possible race conditions as pointed out by @StephanErb, so my vote would be to merge this if everyone's happy with it and close #3256 until someone observes that same behavior again.

**Comment by TomAugspurger** [comment]
Has anyone seen this CI failure before?

```
=================================== FAILURES ===================================

____________________________ test_pickle_functions _____________________________

    def test_pickle_functions():

        def make_closure():

            value = 1

    

            def f(x):  # closure

                return x + value

    

            return f

    

        def funcs():

            yield make_closure()

            yield (lambda x: x + 1)

            yield partial(add, 1)

    

        for func in funcs():

            wr = weakref.ref(func)

            func2 = loads(dumps(func))

            wr2 = weakref.ref(func2)

            assert func2(1) == func(1)

            del func, func2

            gc.collect()

>           assert wr() is None

E           AssertionError: assert <function test_pickle_functions.<locals>.funcs.<locals>.<lambda> at 0x7f0ab6f535f0> is None

E            +  where <function test_pickle_functions.<locals>.funcs.<locals>.<lambda> at 0x7f0ab6f535f0> = <weakref at 0x7f0a8cc05350; to 'function' at 0x7f0ab6f535f0 (<lambda>)>()

distributed/protocol/tests/test_pickle.py:47: AssertionError`
```

I've restarted that build.

**Comment by mrocklin** [comment]
Not recently, but yes, it used to be an intermittent failure.

**Comment by TomAugspurger** [comment]
OK, thanks. We'll merge this and keep an eye out for that failure again.

Thanks @rockwellw!

File: distributed/scheduler.py [hunk 1]
```diff
@@ -271,6 +271,12 @@ def __init__(
 
         self.extra = extra or {}
 
+    def __hash__(self):
+        return hash((self.name, self.host))
+
+    def __eq__(self, other):
+        return type(self) == type(other) and hash(self) == hash(other)
```
Comment (StephanErb) [inline distributed/scheduler.py]: Sorry for being late to the party. There is a bug in here: It is not guaranteed that two WorkerState objects are actually the same even though there hash value is the same. Hashing is not a bijective function.

I think we need to go the extra route here and check for `(self.name, self.host) == (other.name, other.host)`

Comment (StephanErb) [reply distributed/scheduler.py]: @TomAugspurger  @bnaul  please see above.

While I totally agree that hash collisions are very very unlikely, Dask has a massive install base with sometimes hundreds to thousands of workers per cluster. So as far as we know the once in a million chance event be next Tuesday. 🤷‍♂

**Pull request merged**